[MRG+3] edit train/test_size default behavior

nelson-liu opened the pull request:
#### Reference Issue

Fixes #5948, #4618
#### What does this implement/fix? Explain your changes.

Changes the default behavior of `train_size` and `test_size` in splitters.
The defaults for both parameters are now `None`
- if `train_size` and `test_size` are both `None`, `train_size` is set to `0.9` and `test_size` is set to `0.1`
- if only one of `train_size` or `test_size` are set, then the value of the unset parameter is `n_samples - set_parameter` (the complement)
- If both are explicitly set, check that they are equal to or less than 1 or `n_samples` and respect that the user wants to subsample the dataset
#### Any other comments?

I'm aware that the `cross_validation` module is being deprecated; does this mean that I shouldn't add these changes to it and only to `model_selection`?
#### TODO
- [x] Edit documentation to clearly explain these new changes

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/scikit-learn/scikit-learn/7459)

<!-- Reviewable:end -->

nelson-liu committed:
edit train/test_size default behavior

nelson-liu committed:
edit train/test_size default behavior

jnothman commented:
Please don't modify cross_validation.

On 20 September 2016 at 13:35, Nelson Liu <notifications@github.com> wrote:

> Reference Issue
>
> Fixes #5948 <https://github.com/scikit-learn/scikit-learn/issues/5948>
> What does this implement/fix? Explain your changes.
>
> Changes the default behavior of train_size and test_size in splitters.
> The defaults for both parameters are now None
>
>    - if train_size and test_size are both None, train_size is set to 0.1
>    and test_size is set to0.9`
>    - if only one of train_size or test_size are set, then the value of
>    the unset parameter is n_samples - set_parameter (the complement)
>    - If both are explicitly set, check that they are equal to or less
>    than 0 and respect that the user wants to subsample the dataset
>
> Any other comments?
>
> I'm aware that the cross_validation module is being deprecated; does this
> mean that I shouldn't add these changes to it and only to model_selection?
> TODO
>
>    - Edit documentation to clearly explain these new changes
>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/scikit-learn/scikit-learn/pull/7459
> Commit Summary
>
>    - edit train/test_size default behavior
>
> File Changes
>
>    - *M* sklearn/cross_validation.py
>    <https://github.com/scikit-learn/scikit-learn/pull/7459/files#diff-0>
>    (41)
>    - *M* sklearn/model_selection/_split.py
>    <https://github.com/scikit-learn/scikit-learn/pull/7459/files#diff-1>
>    (54)
>    - *M* sklearn/model_selection/tests/test_split.py
>    <https://github.com/scikit-learn/scikit-learn/pull/7459/files#diff-2>
>    (15)
>    - *M* sklearn/tests/test_cross_validation.py
>    <https://github.com/scikit-learn/scikit-learn/pull/7459/files#diff-3>
>    (36)
>
> Patch Links:
>
>    - https://github.com/scikit-learn/scikit-learn/pull/7459.patch
>    - https://github.com/scikit-learn/scikit-learn/pull/7459.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7459>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6w0JaHZaFQeJaKrIaip96r33SnH0ks5qr1SAgaJpZM4KBMM_>
> .
>

nelson-liu committed:
revert changes to cross_Validation

nelson-liu committed:
revert changes to cross_Validation

nelson-liu committed:
fix improper merge resolution

nelson-liu commented:
> Please don't modify cross_validation.

ok, thanks.

lesteve commented:
This needs a rebase on master to fix the conflicts.

nelson-liu commented:
> This needs a rebase on master to fix the conflicts.

Done, thanks for letting me know. Didn't notice / realize someone else was working on the file as well :)

nelson-liu committed:
edit default train/test_size behavior for other splitters

nelson-liu commented:
this is ready to be looked at, missing a whatsnew though (I'd like to verify the current behavior is correct before writing it). Among the changes discussed earlier, `GroupShuffleSplit` has test_size set to a default of `0.2`. This PR changes it to `0.1` to match all the other splitters.

amueller commented:
I expect the docs of ``train_test_split`` also need a change. And you can't change the default behavior of ``GroupShuffleSplit`` without a deprecation. We could have done that in 0.18, but now it's too late :-/

nelson-liu commented:
right, sorry for omitting that @amueller . I've actually never "deprecated" by changing the default parameter of a function before, would you mind pointing me to an example / explaining how to do it? It seems to be missing in the contributing docs as well, so I'd be happy to add it there.

amueller commented:
I'm not sure we should actually change the default from 0.2 to 0.1 in that case.

For reference, the deprecation documentation is here:
http://scikit-learn.org/dev/developers/contributing.html#deprecation

amueller commented:
I'm always kinda confused by the fact that the "Contributing" link on the help page doesn't go to contributing, but to developer....

nelson-liu commented:
hmm, would
```
            warnings.warn("The default value of train_size parameter was "
                          "changed from 0.2 to None, which changes the "
                          "default (train_size, test_size)"
                          "from (0.2, 0.8) to (0.1, 0.9) "
                          "in version 0.19", DeprecationWarning)
```
be clear? I feel like there's a better way to word it...

amueller commented:
Usually we just say "The default value of (..) will change from .. to .. in version 0.20" and the ``None`` is really just a weird implementation detail

nelson-liu commented:
@amueller we also need a deprecation for `train_test_split` right? Considering the defaults will change from `0.25,0.75` to `0.1,0.9` in version 0.21

also, by `the docs of train_test_split need a change, do you mean the docstrings or the examples / tutorial material`?

nelson-liu committed:
add deprecation warnings to groupshufflesplit and train_test_split

nelson-liu commented:
also I edited the default parameters taken by the `__init__` of GroupShuffleSplit to take `test_size=None` and `train_size=None`, allowing me to throw the deprecation warning and then accordingly set the previous defaults (`test_size=0.2` and `train_size-0.8`). Should I just not have touched this?

amueller commented:
We should not change the defaults of ``train_test_split``. We should use the same behavior, though. And I meant the docstrings need to document the new way ``train_size`` and ``test_size`` interact.

jnothman commented:
There are a couple of errors in your issue description:

> * if train_size and test_size are both None, train_size is set to 0.1 and test_size is set to 0.9

you mean the opposite

> * If both are explicitly set, check that they are equal to or less than 0 and respect that the user wants to subsample the dataset

I think you mean less than `n_samples` or less than 1.

Now that `model_selection` is released we need to be aware of backwards compatibility issues. For now, if a user sets `train_size` without setting `test_size` (and `train_size+test_size > 1` or `n_samples`) the best we can do is warn that the behaviour will change in a future release, IMO.

nelson-liu commented:
> We should not change the defaults of train_test_split. We should use the same behavior, though. And I meant the docstrings need to document the new way train_size and test_size interact.

Makes sense, I've reverted the change and updated the description accordingly.

> There are a couple of errors in your issue description:

Indeed, thanks for catching that.

 > Now that model_selection is released we need to be aware of backwards compatibility issues. For now, if a user sets train_size without setting test_size (and train_size+test_size > 1 or n_samples) the best we can do is warn that the behaviour will change in a future release, IMO.

To clarify --- we should not implement the changes to the behavior when setting train_size without test_size, but throw a warning instead saying that it will change in future release (sort of like a deprecation)? Also, how is there a case that `train_size+test_size > 1 or n_samples`? Sorry if these questions have obvious answers.

jnothman commented:
I'm not now sure about the details of my comment. Point is that all behaviours need to remain identical, except perhaps for a warning or *extra* user-specified parameters/values handled. I need to take a proper look at what's changing here to give more precise feedback but there's a huge load of issues for me to skim right now...

jnothman commented:
I hope to look at this soon, @nelson-liu. In the meantime, resolving merge conflicts with master wouldn't hurt.

amueller commented:
@jnothman feel free to hassle me about this, too ;)

lesteve commented:
Rebased on master to fix conflicts.

nelson-liu commented:
not sure if this is ready to be looked at actually, I don't think i've addressed @jnothman 's latest comments wrt keeping behaviors identical...

jnothman commented on sklearn/model_selection/_split.py:
```diff
@@ -994,20 +994,23 @@ class ShuffleSplit(BaseShuffleSplit):
+    test_size : float, int, or None, default None
+        If float, should be between 0.0 and 1.0 and represent the proportion
+        of the dataset to include in the test split. If int, represents the
+        absolute number of test samples. If None, and `train_size` is None,
```
double backticks

jnothman commented on sklearn/model_selection/_split.py:
```diff
@@ -994,20 +994,23 @@ class ShuffleSplit(BaseShuffleSplit):
+        If float, should be between 0.0 and 1.0 and represent the proportion
+        of the dataset to include in the test split. If int, represents the
+        absolute number of test samples. If None, and `train_size` is None,
+        the value is set to 0.1. If None and `train_size` is not None, the
```
double backticks here too

jnothman commented on sklearn/model_selection/_split.py:
```diff
@@ -1079,24 +1083,34 @@ class GroupShuffleSplit(ShuffleSplit):
                  random_state=None):
+        if test_size is None and train_size is None:
+            warnings.warn("The default value of the test_size parameter"
+                          "will change from 0.1 to 0.2 in version 0.21.",
```
0.2 is already the default, is it not??

jnothman reviewed (changes requested):
Unfortunately, we need to retain the behaviour that `train_test_split(train_size=.5)` will sample 10% (or whatever the default is) for test and 50% for train. This behaviour can change in version 0.21. We achieve this by setting `test_size` to some otherwise inappropriate sentinel by default, such as `'default'`, which behaves exactly like the current default value (0.1). In 0.21, it can change to None and behave like you specify. In the case that `test_size='default'` and `train_size` is a number, we warn that behaviour will change to one where `test_size` will always complement `train_size` unless both are specified or both are unspecified.

nelson-liu committed:
correctly format docstrings and remove warnings of changed default values

nelson-liu committed:
restored original behavior with added DeprecationWarnings

nelson-liu committed:
add unit tests for deprecationwarnings

nelson-liu committed:
reset GroupShuffleSplit default test_size to 0.2

nelson-liu committed:
remove extraneous test

nelson-liu commented:
@jnothman - let me know if i understood what to do correctly. The idea is that the current behavior shouldn't change, and thus this PR should just add the functions that detect when a user invokes behavior that will change in the future (`test_size` default, unspecified value, `train_size` some specified value) and warn them that the behavior of the code will change in 0.21. Then when 0.20dev comes along, we'll actually implement the new behavior?

If this is the case, I reverted everything to ensure that it maintains the current behavior and just warns the user appropriately.

nelson-liu commented:
another thing i was thinking about wrt this PR: I set the default value to `'default'`, but should `test_size='default'` or the actual value appear in `__repr__`?

for example, in the current PR, printing a `GroupShuffleSplit` object will show `test_size=0.2` (as opposed to `test_size='default'`), and printing a `StratifiedShuffleSplit` object will show `test_size='default'` (as opposed to `test_size = 'default'`)

nelson-liu commented:
I also reverted any changes I made to any default values --- this PR doesn't touch those at all (e.g. `GroupShuffleSplit` "default" test size is still 0.2, and `StratifiedShuffleSplit` "default" test size is still 0.1)

nelson-liu committed:
fix flake8 violations

nelson-liu committed:
fix indentation error overriding test size in groupsamplesplit

nelson-liu commented:
hmm, the circleci error looks unrelated?

jnothman commented:
I think we conventionally don't touch `__repr__`, annoying as that may be.

If you've not merged/rebased with master recently, that's probably the source of the circle error.

jnothman commented:
I.e. please merge with updated master

jnothman commented on sklearn/model_selection/_split.py:
```diff
@@ -994,16 +994,16 @@ class ShuffleSplit(BaseShuffleSplit):
-        proportion of the dataset to include in the test split. If
-        int, represents the absolute number of test samples. If None,
-        the value is automatically set to the complement of the train size.
+    test_size : float, int, None, or 'default', default 'default'
```
we'd just say "optional" and describe the default behaviour below. We don't really want users explicitly specifying 'default'.

jnothman commented on sklearn/model_selection/_split.py:
```diff
@@ -994,16 +994,16 @@ class ShuffleSplit(BaseShuffleSplit):
+        If float, should be between 0.0 and 1.0 and represent the proportion
+        of the dataset to include in the test split. If int, represents the
+        absolute number of test samples. If None, the value is set to the
+        complement of the train size. If 'default', the value is set to 0.1.
```
"By default".

Then describe 'The default will change in version ... : it will remain 0.1 only if train_size is unspecified, otherwise it will complement. The specified train_size."

jnothman commented on sklearn/model_selection/_split.py:
```diff
@@ -1095,8 +1098,16 @@ class GroupShuffleSplit(ShuffleSplit):
+            if train_size is not None:
+                warnings.warn("test_size will always complement train_size "
+                              "unless both are specified or both are "
+                              "unspecified in version 0.21.",
```
Put "From version 0.21" at the start. Make this a `FutureWarning`?

jnothman commented on sklearn/model_selection/_split.py:
```diff
@@ -1339,7 +1359,8 @@ def _validate_shuffle_split_init(test_size, train_size):
                     'than 1.0 or be an integer' % test_size)
-        elif np.asarray(test_size).dtype.kind != 'i':
+        elif (np.asarray(test_size).dtype.kind != 'i' and
+              test_size != "default"):
```
how can test_size be 'default' here if you replace the value above?

nelson-liu replied:
you're right, thanks for spotting my oversight

nelson-liu committed:
change DeprecationWarnings to FutureWarnings

nelson-liu committed:
reword docstrings for test_size parameters

nelson-liu commented:
thanks again for the review @jnothman

nelson-liu committed:
fix flake8 error in line length

nelson-liu committed:
remove extraneous newline

jnothman commented on sklearn/model_selection/_split.py:
```diff
@@ -1339,7 +1365,7 @@ def _validate_shuffle_split_init(test_size, train_size):
                     'test_size=%f should be smaller '
                     'than 1.0 or be an integer' % test_size)
-        elif np.asarray(test_size).dtype.kind != 'i':
+        elif (np.asarray(test_size).dtype.kind != 'i'):
```
remove the parentheses

jnothman commented on sklearn/model_selection/_split.py:
```diff
@@ -1330,6 +1348,14 @@ def _validate_shuffle_split_init(test_size, train_size):
     """
+    if test_size == "default":
+        if train_size is not None:
+                warnings.warn("From version 0.21, test_size will always "
```
indentation

jnothman commented on sklearn/model_selection/_split.py:
```diff
@@ -1674,14 +1707,22 @@ def train_test_split(*arrays, **options):
+    if test_size == 'default':
+        test_size = None
+        if train_size is not None:
+                warnings.warn("From version 0.21, test_size will always "
```
indentation

jnothman commented on sklearn/model_selection/_split.py:
```diff
@@ -1674,14 +1707,22 @@ def train_test_split(*arrays, **options):
+        if train_size is not None:
+                warnings.warn("From version 0.21, test_size will always "
+                              "complement train_size unless both "
+                              "are specified or both are unspecified.",
```
If both are unspecified, will test_size not complement train_size?

nelson-liu replied:
indeed, thanks for catching that.

nelson-liu committed:
edit indentation errors and clarify future test_size behavior

nelson-liu committed:
ignore FutureWarnings in unrelated tests

nelson-liu committed:
fix flake8 error

jnothman commented:
I think you can change this to MRG.

jnothman commented:
Especially since this LGTM

nelson-liu commented:
great, thanks @jnothman

nelson-liu commented:
reping on this / @jnothman perhaps mark approved on github?

nelson-liu commented:
pinging on this again, since i'd be able to work on it for the next few days if there's anything that needs to be fixed.

jmschrei commented on sklearn/model_selection/_split.py:
```diff
@@ -994,16 +994,19 @@ class ShuffleSplit(BaseShuffleSplit):
-        proportion of the dataset to include in the test split. If
-        int, represents the absolute number of test samples. If None,
-        the value is automatically set to the complement of the train size.
+    test_size : float, int, None, optional
```
Should specify 'default "default"' in the same way the other parameters specify their defaults

jmschrei commented on sklearn/model_selection/_split.py:
```diff
@@ -994,16 +994,19 @@ class ShuffleSplit(BaseShuffleSplit):
+        If float, should be between 0.0 and 1.0 and represent the proportion
+        of the dataset to include in the test split. If int, represents the
+        absolute number of test samples. If None, the value is set to the
+        complement of the train size. By default, the value is set to 0.1.
```
Should explicitly mention what "default" means here.

lesteve replied:
> By default, the value is set to 0.1.

Hmmm really? Quickly looking at the code below, it looks like the default value is 0.2 ...

nelson-liu replied:
whoops, good catch @lesteve .

jmschrei commented:
I can't speak to how development is progressing in other areas of sklearn like others can, but in general these changes LGTM.

raghavrv commented:
This looks nice to me... Can you rebase/merge master... We can merge this soon...

nelson-liu committed:
add more details about the defaults

nelson-liu committed:
Merge branch 'master' into edit_train_test_split_api

raghavrv commented:
And a whatsnew :)

jnothman commented:
Yes, please mention under API changes, then we'll merge.

nelson-liu commented:
i'm not too sure how to word this, but I was thinking something along the lines of: 
``` 
In version 0.21, the default behavior of splitters that use the `test_size` and `train_size` parameter will change. Currently, the default value of `test_size` is `0.1`. Thus, if `train_size` is specified and `test_size` is not, `test_size` is still implicitly `0.1`. In the version 0.21, if `train_size` is specified and `test_size` is not, `test_size` will be instead set to the complement of `train_size`.
```
but this seems a little too narrative? I want to provide some intuition / context for why this change is happening, since I suspect it could affect many.

lesteve commented:
> but this seems a little too narrative?

Maybe you can put a short snippet and mention the values of test_size/train_size in both 0.20 and 0.21. By the way it may be worth creating an issue to make sure we remove the warnings in 0.21 and we add an entry in the 0.21 changelog in the "API changes" section ...

nelson-liu committed:
fix typo in GroupShuffleSplit stating default is 0.1

jnothman commented:
Sufficient what's new:

In version 0.21, the default behavior of splitters that use the `test_size` and `train_size` parameter will change, such that specifying `train_size` alone will cause `test_size` to be the remainder.

jnothman committed:
Add what's new

jnothman committed:
Merge branch 'master' into edit_train_test_split_api

jnothman commented:
Thanks @nelson-liu

nelson-liu commented:
sorry for letting this languish @jnothman , and thanks for taking it into your own hands to patch it up.

jnothman commented:
No worries! I'm glad it is in and we can stop having people frustrated by
it.

On 15 June 2017 at 08:41, Nelson Liu <notifications@github.com> wrote:

> sorry for letting this languish @jnothman <https://github.com/jnothman> ,
> and thanks for taking it into your own hands to patch it up.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7459#issuecomment-308578564>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6_4J-x0YhQAiI_EelADBaO31QJZnks5sEGGrgaJpZM4KBMM_>
> .
>

eliot1785 commented:
So what actually changed here, in practice? If the user did nothing they would have 0.75/0.25 before, and it looks like that is still the case.

nelson-liu commented:
> So what actually changed here, in practice? 

No breaking behavior has been implemented in version `0.19` (hence the `FutureWarning`). In version `0.21`, if you specify `train_size` without specifying `test_size`, it will automatically set `test_size = 1 - train_size`

> If the user did nothing they would have 0.75/0.25 before, and it looks like that is still the case.

Yes, that is correct.

eliot1785 commented:
Right, so what I'm confused by is it seems like that's already the behavior. From the doc: "If None, the value is set to the complement of the train size."

nelson-liu commented:
ah, `test_size` is not `None` by default (it's now a sentinel string `"default"`, and was the float `0.1` in version `0.18`). If you explicitly pass `None` to it, then it will complement the train size. this has always been the case.

Does that help clarify?

eliot1785 commented:
That helps, although I would have thought the default was the float 0.25 since the docs also say, "By default, the value is set to 0.25." Was it 0.1 only in version 0.18?

jnothman commented:
defaults and their documentation were inconsistent in 0.18 iirc.

nelson-liu commented:
Also, note that the defaults are different for different splitters (e.g. `train_test_split` does have a default of `0.25` instead of `0.1`).

eliot1785 commented:
But you're saying it was 0.1 for one version?

nelson-liu commented:
The default `test_size` for `BaseSplitter` has been `0.1` for at least [0.17](https://github.com/scikit-learn/scikit-learn/blob/9e3ebd43e76b8f2f99dbd94afb7e385183c7ea12/sklearn/cross_validation.py#L726) and [0.18](https://github.com/scikit-learn/scikit-learn/blob/ab93d657eb4268ac20c4db01c48065b5a1bfe80d/sklearn/model_selection/_split.py#L920) (and possibly longer).

As far as i can remember, the default `test_size` for `train_test_split` has been `0.25` (it could have been changed from 0.1, though -- you can probably find out by looking at the release history to find out)

In addition: I'm not sure what objects you're referring to by "it" --- the default behavior of the Splitter classes and `train_test_split` are different (and this PR touches both, so unsure which you're referring to). Hopefully I answered your question though.

eliot1785 commented:
Ok thanks. I just wanted to make sure the defaults for train_test_split hadn't changed recently because if so it would affect my models. It doesn't sound like it, so I'm satisfied. I guess this is a good example of the wisdom of always specifying parameters rather than relying on defaults.
